### PR TITLE
Container changes

### PR DESCRIPTION
**Pull request by sachinmsft**
PR for populating container metrics.

### CODE REVIEW

**Comment by sachinmsft** [comment]
CI is giving this error **collector\container.go:9:2: cannot find package "github.com/Microsoft/hcsshim" in any of:**
can we load github.com/Microsoft/hcsshim pkg to machine?

File: collector/container.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	ContainerAvailable *prometheus.Desc
+	// memory
+	UsageCommitBytes            *prometheus.Desc
+	UsageCommitPeakBytes        *prometheus.Desc
+	UsagePrivateWorkingSetBytes *prometheus.Desc
+
+	// CPU
+	TotalRuntime100ns  *prometheus.Desc
+	RuntimeUser100ns   *prometheus.Desc
+	RuntimeKernel100ns *prometheus.Desc
```
Comment (carlpett) [inline collector/container.go]: Since these three are scaled to seconds (right?), remove the 100ns suffix to avoid confusion

File: collector/container.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+			nil,
+		),
+		UsagePrivateWorkingSetBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "memory_usage_private_working_set_bytes"),
+			"Memory Usage Private Working Set Bytes",
+			[]string{"container_id"},
+			nil,
+		),
+		TotalRuntime100ns: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "cpu_usage_seconds_total"),
```
Comment (carlpett) [inline collector/container.go]: Does usermode and kernelmode add upp to total, or is there some "other" time too? It might be more in line with how we report other cpu usage to have a single metric with a `mode` label, if possible.

File: collector/container.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+	// Types Container is passed to get the containers compute systems only
+	containers, err := hcsshim.GetContainers(hcsshim.ComputeSystemQuery{Types: []string{"Container"}})
+	if err != nil {
+		log.Error("Err in Getting containers:", err)
+		return nil, err
+	}
+
+	if len(containers) == 0 {
+		log.Warn("no containers on node")
```
Comment (carlpett) [inline collector/container.go]: I don't think we need to log a warning here. It might however be nice to add a metric for the number of containers on the host?

File: collector/container.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.RuntimeKernel100ns,
+			prometheus.CounterValue,
+			float64(cstats.Processor.RuntimeKernel100ns)*ticksToSecondsScaleFactor,
+			containerId,
+		)
+
+		if len(cstats.Network) == 0 {
+			log.Warn("No Network Stats for container: ", containerId)
```
Comment (carlpett) [inline collector/container.go]: Does this indicate an invalid container configuration? Otherwise, like the warning for no containers above, I'd skip the warning and just possibly add a metric for the number of interfaces on the container.

File: collector/container.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+			continue
+		}
+
+		networkStats := cstats.Network
+
+		for _, interafce := range networkStats {
```
Comment (carlpett) [inline collector/container.go]: Typo, `interafce`->`interface`

File: collector/container.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
+// returns containers metrics from HCS
```
Comment (carlpett) [inline collector/container.go]: This becomes a package-level comment if viewed in godoc, which is confusing. If we want to have this comment at all (I'm not sure it adds much information that isn't readily apparent from skimming the code), merge it with the struct comment for `ContainerMetricsCollector`

**Comment by carlpett** [comment]
>CI is giving this error **collector\container.go:9:2: cannot find package "github.com/Microsoft/hcsshim" in any of:**
>can we load github.com/Microsoft/hcsshim pkg to machine?

We're using vendoring our dependencies with `dep`, so you need to add it by running `dep ensure -add github.com/microsoft/hcsshim`.

File: collector/container.go (hunk 4 above)
Comment (sachinmsft) [reply collector/container.go]: Yes. There should be network stats in normal scenario so having this log should be okay. I changed it to info instead of warning.

File: collector/container.go (hunk 3 above)
Comment (sachinmsft) [reply collector/container.go]: Done.

File: collector/container.go (hunk 2 above)
Comment (sachinmsft) [reply collector/container.go]: I have to confirm with the folks who wrote the hcsshim layer. Though as we have only 2 modes so having separate metrics for them should not be big problem.

File: collector/container.go (hunk 1 above)
Comment (sachinmsft) [reply collector/container.go]: corrected.

File: collector/container.go (hunk 5 above)
Comment (sachinmsft) [reply collector/container.go]: corrected.

File: collector/container.go (hunk 6 above)
Comment (sachinmsft) [reply collector/container.go]: Removed it.

**Comment by sachinmsft** [comment]
> > CI is giving this error **collector\container.go:9:2: cannot find package "github.com/Microsoft/hcsshim" in any of:**
> > can we load github.com/Microsoft/hcsshim pkg to machine?
> 
> We're using vendoring our dependencies with `dep`, so you need to add it by running `dep ensure -add github.com/microsoft/hcsshim`.

I added the dependency in MakeFile, but still CI is not passing. Am I making any mistake. ?

**Comment by carlpett** [comment]
Ah, sorry about being unclear regarding vendoring. You seem to have fixed it since though (not adding dep ensure in Makefile, but checking in the vendor folder), good!
Currently the build seems to fail on some CGO compilation, but only while compiling the test code. This is odd, since we are pruning out test code, so vendoring shouldn't affect test...

Could you verify your version of  `dep`? It should be at least 0.4.1 (but there's no reason not to use 0.5.0).

**Comment by sachinmsft** [comment]
I am using the correct version of dep. Actually Microsoft/hcsshim has dependency on runtime/cgo.
I think during make test it try to run the test and runtime/cgo serach for C compiler. I tried to install min-gw but somehow it is not working properly. 
I was able to run locally by installing TDM-GCC compiler(http://tdm-gcc.tdragon.net/download), but his pkg is not available through choco.
is it possible to configure the appveyor machine manually and install this application?

**Comment by sachinmsft** [comment]
@carlpett, Do you have any suggestions on this. BTW, Is it possible to not build the vendor code on make test.?

**Comment by carlpett** [comment]
Sorry for being slow to reply, been busy with dayjob stuff lately.
As I mentioned, `dep` _should_ be pruning out all the vendor testing code, so I don't see why this doesn't seem to happen here. You could try `dep prune` to run another pass over it, I guess, although it is supposed to be obsolete.

We can do some extra configuration on the machine if necessary, although VS2017 is already installed on the machine, which I would expect to be enough?

**Comment by sachinmsft** [comment]
the compiler was already installed on machine. just path was missing. It worked finally.

**Comment by sachinmsft** [comment]
@carlpett due to https://github.com/martinlindhe/wmi_exporter/issues/321 I assume you are waiting till the end of the week to merge this.
I have made a change to consume these container metrics in the kubernetes-mixin project https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/164 . They do not want to merge that PR until there is a wmi_exporter release that contains these container changes. 
Is there anything I can do to help to speed up the time for a release once this is merged?

File: collector/container.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+	// Types Container is passed to get the containers compute systems only
+	containers, err := hcsshim.GetContainers(hcsshim.ComputeSystemQuery{Types: []string{"Container"}})
+	if err != nil {
+		log.Error("Err in Getting containers:", err)
+		return nil, err
+	}
+
+	count := len(containers)
+
+	if count == 0 {
```
Comment (carlpett) [inline collector/container.go]: We should still be reporting the `ContainersCount` metric here, I think, so move this down below that metric being sent.

File: collector/container.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+		container, err := hcsshim.OpenContainer(containerId)
+		if err != nil {
+			log.Error("err in opening container: ", containerId)
+			continue
+		}
+
+		cstats, err := container.Statistics()
+		if err != nil {
+			log.Error("err in fetching container Statistics: ", containerId)
```
Comment (carlpett) [inline collector/container.go]: I think we want the error logged as well

**Comment by carlpett** [comment]
No worries, I'll make a release as soon as this is merged :)

File: collector/container.go (hunk 8 above)
Comment (sachinmsft) [reply collector/container.go]: added err.

File: collector/container.go (hunk 7 above)
Comment (sachinmsft) [reply collector/container.go]: done

File: collector/container.go [hunk 9]
```diff
@@ -139,6 +139,14 @@ func (c *ContainerMetricsCollector) Collect(ch chan<- prometheus.Metric) error {
 	return nil
 }
 
+// containerClose closes the container resource
+func containerClose(c hcsshim.Container) {
+	err := c.Close()
+	if err != nil {
+		log.Fatal(err)
```
Comment (carlpett) [inline collector/container.go]: We do not want to quit the exporter if this happens, so I'd say a `log.Error` is sufficient.

Comment (sachinmsft) [reply collector/container.go]: Okay. changed to log.Error.

**Comment by sachinmsft** [comment]
go-winio is added as hcsshim has dependecy on this pkg. I am not sure if we can directly delete test tar files and push the changes to vendor. I can open a issue for this and can take it later if this is not blocker.

**Comment by sachinmsft** [comment]
I deleted testdata tar balls.

**Comment by carlpett** [comment]
Great work! Really happy we'll have this capability in the exporter! :)

Just realized that there is one small thing missing though - docs. So sorry I didn't think about that until now! There is a template [here](https://github.com/martinlindhe/wmi_exporter/blob/master/docs/collector-template.md). The `Classes` bit in the table doesn't make much sense here, so feel free to remove that row.

Apart from that, I think this is ready to merge :tada:

**Comment by sachinmsft** [comment]
@carlpett I have added the doc, though build is failing while installing make
choco install gitversion.portable make -y

**Comment by sachinmsft** [comment]
I have fixed the appveyor.yml and build is green now.

File: appveyor.yml [hunk 10]
```diff
@@ -14,7 +14,8 @@ install:
   - set PATH=%PATH%;C:\mingw-w64\x86_64-7.2.0-posix-seh-rt_v5-rev1\mingw64\bin
   - go get -u github.com/prometheus/promu
   - go get -u github.com/alecthomas/gometalinter && gometalinter --install
-  - choco install gitversion.portable make -y
+  - choco install gitversion.portable --pre -y
```
Comment (carlpett) [inline appveyor.yml]: Why is this necessary? I don't think we want prereleases?

**Comment by sachinmsft** [comment]
Build was failing if --pre was not used.
Official page also mentions to use --pre.

https://chocolatey.org/packages/GitVersion.Portable

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by carlpett** [comment]
Only to use that particular beta version. The latest stable release _should_ work, I would think: https://chocolatey.org/packages/GitVersion.Portable/4.0.0.
I tested re-running an older version, and there it installed fine. Perhaps you were a bit unlucky and the package was unavailable for a short period of time? Choco isn't always completely reliable, sadly.

**Comment by carlpett** [comment]
I reverted that last commit, and it works again, so let's get this merged! :smile:

**Comment by carlpett** [comment]
Thanks for this contribution! And for staying with me during the rather long review process, sorry it took so long this time.
I'll make a release as well, should be up in a few minutes!

**Comment by sachinmsft** [comment]
Thanks @carlpett for all the help. I see new release with my changes though it does not have .msi and .exe files to download, it just has source code zipped file. does it take some time to populate executable files.?

**Comment by carlpett** [comment]
Interestingly, the build failed on _another_ chocolatey package. Retriggered the build and all went well, binaries are now uploaded!

**Comment by sachinmsft** [comment]
Yes. Thanks a lot.

**Comment by szediktam** [comment]
@sachinmsft  container metrics can have the namespace/pod/container label?

**Comment by sachinmsft** [comment]
> @sachinmsft container metrics can have the namespace/pod/container label?

container metrics are retrieved directly from hcs. which is not aware of pod/container labels. it just have container Id.
